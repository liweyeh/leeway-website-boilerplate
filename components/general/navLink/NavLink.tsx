/**
 * Navigation link component for internal routing
 * @author Liwei Yeh <liwei.yeh@adasasistemas.com>
 */
// Library
import { useState, useRef } from 'react';
import { useRouter } from 'next/router';
import Link from 'next/link';
import { isEmpty } from 'lodash';

// Components
import { Modal, Button } from '@components';

// CSS Style for this module
import styles from './navLink.module.scss';

/**
 * Interface for NavLink props
 */
interface NavLinkProps {
	/**
	 * Display name or element
	 */
	children: React.ReactNode;
	/**
	 * External link
	 */
	isExternal?: boolean;
	/**
	 * On click
	 */
	onClick?: (e: React.MouseEvent<HTMLAnchorElement>) => void;
	/**
	 * Desitnation of the anchor
	 */
	href: string;
}
/**
 * The NavLink component allows you to easily customize anchor elements
 * @param {NavLinkProps} Prop
 * @returns  a styled NavLink
 */
export const NavLink = ({
	href,
	children,
	isExternal = false,
	onClick = () => {
		return;
	},
}: NavLinkProps) => {
	const router = useRouter();
	const isActive = !isEmpty(router) && router.pathname === href;
	const linkRef = useRef<HTMLAnchorElement>(null);
	const [openModal, setOpenModal] = useState<boolean>(false);

	const clickHandle = () => {
		linkRef?.current?.click();
	};

	return (
		<>
			{isActive && <span className={styles['visually-hidden']}>Current page: </span>}
			{isExternal ? (
				<a
					aria-current={isActive}
					onClick={() => setOpenModal(true)}
					target="_blank"
					rel="noreferrer"
					className={styles['link']}
				>
					{children}
				</a>
			) : (
				<Link href={href} onClick={onClick}>
					{children}
				</Link>
			)}
			<Modal
				open={openModal}
				handleClose={() => setOpenModal(false)}
				customModalStyle={styles['nav-link-modal']}
				customBackdropStyle={styles['nav-link-modal-backdrop']}
			>
				<h4>You are about to exit the website.</h4>
				<h4>Are you sure ?</h4>
				<div className={styles['modal-buttons']}>
					<Button customStyles={styles['modal-button']} onClick={() => setOpenModal(false)}>
						<p>No</p>
					</Button>
					<Button onClick={clickHandle} customStyles={styles['modal-button']}>
						<a
							href={href}
							ref={linkRef}
							target="_blank"
							rel="noreferrer"
							className={styles['link']}
							onClick={onClick}
						>
							<p>Yes</p>
						</a>
					</Button>
				</div>
			</Modal>
		</>
	);
};
