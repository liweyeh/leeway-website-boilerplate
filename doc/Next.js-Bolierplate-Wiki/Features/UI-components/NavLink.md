[[_TOC_]]

# Overview

- Purpose: A styled anchor element with proper accessbility support
- Type: General UI
- Status: ðŸš€ **Completed**
- Code: [URL](https://dev.azure.com/ADASA-Accelerator/Website-React-Booster/_git/website-boilerplate-nextjs?path=/components/general/navLink)
- Created by: Liwei Yeh <liwei.yeh@adasasistemas.com>
- Created time: 25/08/2022

# Props

\* = required

- children<`React.ReactNode`>\*: Display name or element
- href<`string`>\*: Desitnation of the anchor
- isExternal<`boolean`>: Dictate whether the link is external, which come with another behaviour
- onClick<`string`>: Callback function when link is clicked

# Usage

```jsx
<NavLink
	href={'https://www.google.com'}
	isExternal={true}
	onClick={() => {
		setIsOpen(false);
	}}
>
	<p>{displayName}</p>
</NavLink>
```
