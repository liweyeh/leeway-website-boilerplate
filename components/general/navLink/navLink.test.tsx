import { render } from '@testing-library/react';
import { NavLink } from './NavLink';
import '@testing-library/jest-dom';

const setup = () => {
	const { asFragment } = render(
		<NavLink href="www.google.com">
			<p>hi</p>
		</NavLink>,
	);
	return { asFragment };
};

test('test snapshot', () => {
	const { asFragment } = setup();
	expect(asFragment()).toMatchSnapshot();
});
