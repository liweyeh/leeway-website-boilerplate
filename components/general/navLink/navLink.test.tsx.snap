// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshot 1`] = `
<DocumentFragment>
  <a
    aria-current="false"
    class="link"
    href="/www.google.com"
  >
    <p>
      hi
    </p>
  </a>
</DocumentFragment>
`;
